llama_index/llms/llama_cpp.py

```diff
@@ -173,21 +173,24 @@ def _download_url(self, model_url: str, model_path: str) -> None:
     @llm_chat_callback()
     def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
         prompt = self.messages_to_prompt(messages)
-        completion_response = self.complete(prompt, **kwargs)
+        completion_response = self.complete(prompt, formatted=True, **kwargs)
```

Comment:
why.... guys, just, WHY

Reply:
@piotrmasior what is the issue here?

Reply:
can you study complete and stream_complete definitions ? there is IF there .... and part of the code that will never be invocated

was struggling with it in part of the privateGPT investigation why my method is never called.... did some comparisions in your source code, with different implementations, guess what, noone ever figured out to actively in the code, filter out kwargs

Reply:
What, this?

```
is_formatted = kwargs.pop("formatted", False)
if not is_formatted:
    prompt = self.completion_to_prompt(prompt)
```

If you call `chat()` or `stream_chat()` then it will not enter this conditional

If you call `complete()` or `stream_complete()` (without `formatted=True` in kwargs) then it will enter the conditional.

No dead code

Reply:
We dont want to call/stack messages_to_prompt() and completion_to_prompt() on top of each other, since they are different functions with different logic

Reply:
> We dont want to call/stack messages_to_prompt() and completion_to_prompt() on top of each other, since they are different functions with different logic

completely didnt get your answer at that part, can you elaborate it a little, as I am missing probably something 
underhood ? Big possibility someone in privateGPT also implemented something very off the rail just reading the code. 

What messeages_to_prompt have to do with completion_to_prompt in this particular case? I can change that logic in the fly, remove that sick IF, or what is more important, remove that formatted=True and just use top to bottom pass of the arguments with **kwargs. so I am asking about what steering here you are thinking of as I might not be so familiar with this projects and it guts.

but given I found exact MR/PR that introduced that change, and fact it was about monsterAPI and probably in my old head red lamp was just flashing over the horizon, why the hell someone changes it for particular integration, maybe I am old, maybe I am wrong. Tell me son, what the hell you are talking about. How this change:
https://user-images.githubusercontent.com/1107041/285324461-0d103e73-29f3-49c4-841e-3b875dc3fa1f.png
secured execution "on top of each other"

Reply:
why it is not abstracted in to strategy above, when actually https://github.com/run-llama/llama_index/blob/main/llama_index/llms/ollama.py#L124 have very same implementation without this overshadowing argument ? Probably someone was thinking what is he doing.

lets see https://github.com/search?q=repo%3Arun-llama%2Fllama_index+formatted&type=code&p=3

https://github.com/run-llama/llama_index/blob/b492b1d748f4c8119351ab5720082d0d698a6f00/llama_index/llms/huggingface.py#L257

totally different impelementation, and having two ifs, interesting as I guess someone was trying to copy approach not understanding what is he doing.

and finally

https://github.com/run-llama/llama_index/blob/b492b1d748f4c8119351ab5720082d0d698a6f00/llama_index/llms/monsterapi.py#L127

I have to admit, someone was not really sure if passed that formatted with true so many times, IT HAD TO BE ALSO HERE

```# Merge additional options into kwargs
    kwargs.update(additional_options)
```

I rest my case here

Reply:
To me that looks like an error in `ollama`

There's two methods here, `completion_to_prompt()` and `messages_to_prompt()`

For an example of what these methods are supposed to do, check out these two util functions made for Llama2
https://github.com/run-llama/llama_index/blob/b492b1d748f4c8119351ab5720082d0d698a6f00/llama_index/llms/llama_utils.py#L16

So `messages_to_prompt()` converts a list of `ChatMessage` objects to an input appropriate for the LLM.

`completion_to_prompt()` does the same, but this time a string is the input

If you call `messages_to_prompt()`, you need to ensure that it's not formatted again by `completion_to_prompt()`, otherwise, the input to the LLM will not be formatted correctly 

I'm not really sure what the issue is with private-gpt, but this is working as expected for llama-cpp. Ollama seems to have a bug that you pointed out.